mm2src/coins/tendermint/tendermint_coin.rs

```diff
@@ -0,0 +1,660 @@
+    RpcClientInitError(String),
+    InvalidChainId(String),
+    InvalidDenom(String),
+    RpcError(String),
+}
+
+#[derive(Display)]
+enum TendermintCoinRpcError {
+    Prost(prost::DecodeError),
+    InvalidResponse(String),
+    PerformError(String),
+}
+
+impl From<prost::DecodeError> for TendermintCoinRpcError {
+    fn from(err: DecodeError) -> Self { TendermintCoinRpcError::Prost(err) }
+}
+
+impl From<TendermintCoinRpcError> for WithdrawError {
+    fn from(err: TendermintCoinRpcError) -> Self { WithdrawError::Transport(err.to_string()) }
+}
+
+impl From<TendermintCoinRpcError> for BalanceError {
+    fn from(err: TendermintCoinRpcError) -> Self {
+        match err {
+            TendermintCoinRpcError::InvalidResponse(e) => BalanceError::InvalidResponse(e),
+            TendermintCoinRpcError::Prost(e) => BalanceError::InvalidResponse(e.to_string()),
+            TendermintCoinRpcError::PerformError(e) => BalanceError::Transport(e),
+        }
+    }
+}
+
+#[cfg(not(target_arch = "wasm32"))]
+impl From<cosmrs::rpc::Error> for TendermintCoinRpcError {
+    fn from(err: cosmrs::rpc::Error) -> Self { TendermintCoinRpcError::PerformError(err.to_string()) }
+}
+
+#[cfg(target_arch = "wasm32")]
+impl From<PerformError> for TendermintCoinRpcError {
+    fn from(err: PerformError) -> Self { TendermintCoinRpcError::PerformError(err.to_string()) }
+}
+
+fn account_id_from_privkey(priv_key: &[u8], prefix: &str) -> MmResult<AccountId, TendermintInitErrorKind> {
+    let signing_key =
+        SigningKey::from_bytes(priv_key).map_to_mm(|e| TendermintInitErrorKind::InvalidPrivKey(e.to_string()))?;
+
+    signing_key
+        .public_key()
+        .account_id(prefix)
+        .map_to_mm(|e| TendermintInitErrorKind::CouldNotGenerateAccountId(e.to_string()))
+}
+
+fn upper_hex(bytes: &[u8]) -> String {
+    let mut str = hex::encode(bytes);
+    str.make_ascii_uppercase();
+    str
+}
+
+impl TendermintCoin {
+    pub async fn init(
+        ticker: String,
+        protocol_info: TendermintProtocolInfo,
+        activation_params: TendermintActivationParams,
+        priv_key: &[u8],
+    ) -> MmResult<Self, TendermintInitError> {
+        if activation_params.rpc_urls.is_empty() {
+            return MmError::err(TendermintInitError {
+                ticker,
+                kind: TendermintInitErrorKind::EmptyRpcUrls,
+            });
+        }
+
+        let account_id =
+            account_id_from_privkey(priv_key, &protocol_info.account_prefix).mm_err(|kind| TendermintInitError {
+                ticker: ticker.clone(),
+                kind,
+            })?;
+
+        let rpc_client =
+            HttpClient::new(activation_params.rpc_urls[0].as_str()).map_to_mm(|e| TendermintInitError {
+                ticker: ticker.clone(),
+                kind: TendermintInitErrorKind::RpcClientInitError(e.to_string()),
+            })?;
+
+        let chain_id = ChainId::try_from(protocol_info.chain_id).map_to_mm(|e| TendermintInitError {
```

Comment:
`ChainId` seems to [implement Deserialize](https://github.com/informalsystems/tendermint-rs/blob/main/tendermint/src/chain/id.rs#L130-L135). `chain_id: String` can be replaced with `chain_id: ChainId`.

Reply:
Yeah, I've seen it, but decided to leave as is because it allows returning a more specific error 🙂